docs/design/tools/illink/feature-checks.md

```diff
@@ -0,0 +1,281 @@
+```csharp
+if (Feature.IsSupported)
+    Feature.Run();
+```
+
+```csharp
+if (!Feature.IsSupported)
+{
+    // ...
+}
+else
+{
+    Feature.Run();
+}
+```
+
+```csharp
+if (OtherCondition)
+{
+}
+else if (Feature.IsSupported)
+{
+    Feature.Run();
+}
+```
+
+### Short-circuiting boolean guard
+```csharp
+var a = Feature.IsSupported && Feature.Run();
+```
+
+```csharp
+var a = !Feature.IsSupported || Feature.Run();
+```
+
+### Ternary operator guard
+```csharp
+_ = Feature.IsSupported ? Feature.Run() : true;
+```
+
+```csharp
+_ = !Feature.IsSupported ? true : Feature.Run();
+```
+
+### Early throw
+
+```csharp
+if (!Feature.IsSupported)
+    throw new Exception ();
+
+Feature.Run();
+```
+
+### Early return
+
+```csharp
+if (!Feature.IsSupported)
+    return;
+
+Featuer.Run ();
+```
+
+## Unsupported guard patterns
+
+The following guard patterns are not supported by all of the tools. Using one of these patterns will result in warnings being produced by ILLink or ILCompiler. If you need to use one of these, it is safe to suppress the warning by using `UnconditionalSuppressMessageAttribute` as long as the call to the annotated feature is unreachable at runtime, but it is recommended to use one of the supported guard patterns instead.
+
+### Boolean and/or
+
+The short-circuiting boolean operators create branches in the control-flow graph that don't always look straightforward in IL. Sometimes (in Debug mode) these compile into a multiple branches that compute a temporary local which stores the computed condition and is used for the conditional branch. ILLink/ILCompiler aren't guaranteed to do branch removal based on constant propagation of this temporary value, so the following patterns aren't supported guards:
+
+```csharp
+if (Feature.IsSupported && OtherCondition)
+    Feature.Run(); // OK in analyzer, may warn in ILLink/ILCompiler
+```
+
+```csharp
+if (OtherCondition && Feature.IsSupported)
+    Feature.Run(); // OK in analyzer, may warn in ILLink/ILCompiler
+```
+
+```csharp
+if (!Feature.IsSupported || OtherCondition)
+    return;
+
+Feature.Run(); // OK in analyzer, may warn in ILLink/ILCompiler
+```
+
+```csharp
+if (OtherCondition || !Feature.IsSupported)
+    return;
+
+Feature.Run(); // OK in analyzer, may warn in ILLink/ILCompiler
+```
+
+### DoesNotReturnIfAttribute
+`DoesNotReturnIfAttribute` influences the analysis done by the Roslyn analyzer, allowing the use of `Debug.Assert` for example.
+
+```csharp
+Debug.Assert (Feature.IsSupported);
+
+Feature.Run(); // OK in analyzer, warns in ILLLink/ILCompiler
+```
+
+However, ILLink and ILCompiler don't optimize away branches based on `DoesNotReturnIfAttribute`, because the attribute doesn't have runtime semantics, and an incorrectly-applied attribute could result in the guarded branch being reachable at runtime. Or, in the case of `Debug.Assert`, if the assert is used incorrectly, the guarded code may still be reachable at runtime in `Release` builds.
+
+Here are some more examples:
```

Comment:
I think the recommendation should be to not rely on these attributes to avoid trim/AOT warnings - we want to recommend patterns which work the same everywhere.